roscala/src/main/scala/coop/rchain/rosette/Queue.scala

```diff
@@ -0,0 +1,106 @@
+package coop.rchain.rosette
+
+import coop.rchain.rosette.Ob.ABSENT
+
+case class Queue(elems: Tuple) extends Ob {
+
+  def head: Ob = elems.elem.head
+
+  def tail: Ob = elems.elem.last
+
+  def nElems: Ob = Fixnum(elems.elem.length)
+
+  def depth(): Int =
+    elems.elem.length
+
+  def isEmpty(): Boolean =
+    (nElems.asInstanceOf[Fixnum] == Fixnum(0)).value
+
+  def enqueue(ele: Ob): Queue =
+    Queue(Tuple(elems.elem :+ ele))
+
+  def dequeue(): (Ob, Queue) =
+    if (isEmpty()) {
+      (ABSENT, this)
+    } else {
+      val v = elems.elem.head
+      (v, Queue(Tuple(elems.elem.drop(1))))
+    }
+
+  def reset(): Queue =
+    Queue.create()
+
+  override def indexedSize: Ob = nElems
+
+  override def nth(i: Int): Option[Ob] = {
+    val index = i % elems.elem.length
+    elems.elem.lift(index)
+  }
+
+  override def setNth(i: Int, v: Ob): Option[Ob] =
```

Comment:
Why do you return `Option[Ob]`? It should return a new `Queue`.

Reply:
Same as above , but i think it can be Option[Queue] here.